backend/src/Database.hs

```diff
@@ -1,10 +1,18 @@
 module Database where
 
+import Control.Exception (throwIO)
 import Data.Pool (withResource)
-import Database.SQLite.Simple (Only (..), execute, execute_, query, query_)
+import Database.SQLite.Simple (Only (..), execute_, query, query_)
 import Types.App (AppM, Env (..), log)
 import Types.DataField (PlayerId, PlayerName)
-import Types.Database (ReadPlayer (..), WritePlayer (..), WriteProcessedGameReport)
+import Types.Database (ReadPlayer (..), ReadProcessedGameReport (..), WritePlayer (..), WriteProcessedGameReport (..))
+
+data UnexpectedResultException = UnexpectedResultException deriving (Show)
+
+instance Exception UnexpectedResultException
+
+readSingle :: IO [Only r] -> IO r
```

Comment:
Abstracting out a pattern here that I was starting to see in a few places.

Sometimes we want to read a single value from the database. Say, the `id` of a player. The query function always returns `IO [r]` where `r` is whatever your "row" type is. So if you're reading a `MyProductType` you get `IO [MyProductType]`, if you're reading only one field, you get an `IO [Only MyField]` (where Only is the single-element tuple), etc.

Unfortunately this is rather weakly typed (by the admittedly very high bar of our pseudo-dependently typed aspirations) and we ALWAYS get a list back, even if we are expecting a single result.*

So this helper is intended to be used when you know you are expecting one **and only one** thing, and you want to extract it, and throw an exception if anything else happens.

*Aside: what would be cool/hard for this library is type-level queries from which are returned some sized-list type who's size depends on the query :shrug: Guess we can't ask for the universe when some languages out there can't cogently express `this`.

Reply:
Solid addition, this makes reading `tableCount`/`pid`/`winner`/`loser` very tidy looking! It's nice to have the extraction logic abstracted into one place, and throwing an exception in this case makes perfect sense to me.

> Guess we can't ask for the universe when some languages out there can't cogently express `this`

Dead 😂 💀 poor javascript out there just doing its best